File: support_tools/go/upd_commander_checker/internal/checker/classifier.go
package checker

import (
	"path/filepath"
	"strings"
)

var layerNames = map[string]bool{"ui": true, "process": true, "data": true}
var roleNames = map[string]bool{"commander": true, "messenger": true, "processing": true, "compresser": true}
var appRoots = map[string]bool{"app": true, "apps": true, "application": true, "applications": true, "feature": true, "features": true}

func ClassifyPath(path string) ModuleInfo {
	return ClassifyPathWithCommonRoots(path, []string{"common", "shared"})
}

func ClassifyPathWithCommonRoots(path string, commonRoots []string) ModuleInfo {
	directories := pathDirectories(path)
	scope := applicationScope(directories)
	stem := fileStem(path)
	return ModuleInfo{Path: path, Layer: findLayer(scope, commonRoots), Role: findPathRole(scope, stem), ApplicationID: findApplication(directories)}
}

func ClassifyImport(name string) ModuleInfo {
	return ClassifyImportWithCommonRoots(name, []string{"common", "shared"})
}

func ClassifyImportWithCommonRoots(name string, commonRoots []string) ModuleInfo {
	parts := splitParts(name)
	scope := applicationScope(parts)
	return ModuleInfo{Path: name, Layer: findLayer(scope, commonRoots), Role: findRole(scope), ApplicationID: findApplication(parts)}
}

func pathDirectories(value string) []string {
	normalized := strings.ToLower(strings.ReplaceAll(value, "\\", "/"))
	directory := filepath.ToSlash(filepath.Dir(normalized))
	if directory == "." {
		return nil
	}
	return strings.FieldsFunc(directory, func(r rune) bool { return r == '/' })
}

func fileStem(value string) string {
	name := filepath.Base(strings.ReplaceAll(value, "\\", "/"))
	extension := filepath.Ext(name)
	return strings.ToLower(strings.TrimSuffix(name, extension))
}

func splitParts(value string) []string {
	replacer := strings.NewReplacer("\\", "/", ".", "/", "-", "_", "_", "/")
	normalized := strings.ToLower(replacer.Replace(value))
	return strings.FieldsFunc(normalized, func(r rune) bool { return r == '/' })
}

func findName(parts []string, candidates map[string]bool) string {
	for index := len(parts) - 1; index >= 0; index-- {
		if candidates[parts[index]] {
			return parts[index]
		}
	}
	return ""
}

func findPathRole(directories []string, stem string) string {
	if role := findName(directories, roleNames); role != "" {
		return role
	}
	for role := range roleNames {
		if stem == role || strings.HasSuffix(stem, "_"+role) {
			return role
		}
	}
	return ""
}

func findRole(parts []string) string {
	for index := len(parts) - 1; index >= 0; index-- {
		part := parts[index]
		if roleNames[part] {
			return part
		}
		if strings.HasSuffix(part, "commander") {
			return "commander"
		}
		if strings.HasSuffix(part, "messenger") {
			return "messenger"
		}
		if strings.HasSuffix(part, "processing") {
			return "processing"
		}
		if strings.HasSuffix(part, "compresser") {
			return "compresser"
		}
	}
	return ""
}

func findApplication(parts []string) string {
	for index := len(parts) - 2; index >= 0; index-- {
		if appRoots[parts[index]] {
			return parts[index+1]
		}
	}
	return ""
}

func applicationScope(parts []string) []string {
	for index := len(parts) - 2; index >= 0; index-- {
		if appRoots[parts[index]] {
			return parts[index+2:]
		}
	}
	return parts
}

func findLayer(parts []string, commonRoots []string) string {
	common := map[string]bool{}
	for _, root := range commonRoots {
		common[strings.ToLower(root)] = true
	}
	for index := len(parts) - 1; index >= 0; index-- {
		part := parts[index]
		if layerNames[part] {
			return part
		}
		if common[part] {
			return "common"
		}
	}
	return ""
}

File: support_tools/python/upd_commander_checker/src/upd_commander_checker/classifier.py
from collections.abc import Iterable
from pathlib import Path

from .models import ModuleInfo


_LAYER_NAMES = {"ui", "process", "data"}
_ROLE_NAMES = {"commander", "messenger", "processing", "compresser"}
_APPLICATION_MARKERS = {"app", "apps", "application", "applications", "feature", "features"}
_DEFAULT_COMMON_ROOTS = ("common", "shared")


def classify_module(
    path: Path,
    root: Path | None = None,
    common_roots: tuple[str, ...] = _DEFAULT_COMMON_ROOTS,
) -> ModuleInfo:
    module_path = _relative_module_path(path, root)
    directory_parts = [part.lower() for part in module_path.parent.parts]
    stem = module_path.stem.lower().replace("-", "_")
    scoped_parts = _application_scope(directory_parts)

    layer = _find_layer(scoped_parts, common_roots)
    role = _find_path_role(scoped_parts, stem)
    application = _find_application(directory_parts)
    return ModuleInfo(path=path, layer=layer, role=role, application=application)


def classify_import(
    module_name: str,
    common_roots: tuple[str, ...] = _DEFAULT_COMMON_ROOTS,
) -> ModuleInfo:
    path_parts = module_name.lower().replace("-", "_").split(".")
    scoped_parts = _application_scope(path_parts)
    token_parts: list[str] = []
    for part in scoped_parts:
        token_parts.extend(part.split("_"))

    layer = _find_layer(token_parts, common_roots)
    role = _find_role(list(reversed(token_parts)))
    application = _find_application(path_parts)
    return ModuleInfo(
        path=Path(module_name),
        layer=layer,
        role=role,
        application=application,
    )


def _relative_module_path(path: Path, root: Path | None) -> Path:
    if root is None:
        return path
    try:
        return path.relative_to(root)
    except ValueError:
        return path


def _find_name(parts: Iterable[str], candidates: set[str]) -> str | None:
    for part in parts:
        if part in candidates:
            return part
    return None


def _find_path_role(directory_parts: list[str], stem: str) -> str | None:
    directory_role = _find_name(reversed(directory_parts), _ROLE_NAMES)
    if directory_role:
        return directory_role
    for role in _ROLE_NAMES:
        if stem == role or stem.endswith("_" + role):
            return role
    return None


def _find_role(parts: list[str]) -> str | None:
    for part in parts:
        if part in _ROLE_NAMES:
            return part
        if part.endswith("commander"):
            return "commander"
        if part.endswith("messenger"):
            return "messenger"
        if part.endswith("processing"):
            return "processing"
        if part.endswith("compresser"):
            return "compresser"
    return None


def _find_application(parts: list[str]) -> str | None:
    for index in range(len(parts) - 2, -1, -1):
        if parts[index] in _APPLICATION_MARKERS:
            return parts[index + 1]
    return None


def _application_scope(parts: list[str]) -> list[str]:
    for index in range(len(parts) - 2, -1, -1):
        if parts[index] in _APPLICATION_MARKERS:
            return parts[index + 2 :]
    return parts


def _find_layer(parts: list[str], common_roots: tuple[str, ...]) -> str | None:
    common = set(common_roots)
    for part in reversed(parts):
        if part in _LAYER_NAMES:
            return part
        if part in common:
            return "common"
    return None
